### PR TITLE
test(controller): testing controllers

### PR DESCRIPTION
**Pull request by SuphawitCE**
# What are changes?
- Update `README.md` to include a prerequisite section
- Test auth middleware
- Organizing tests
- Mock function with stub
- Testing asynchronous code
- Using hooks `before` and `after`
- Refactor test

# Related documentation
Mocha: https://mochajs.org/
Chai: https://www.chaijs.com/
Sinon: https://sinonjs.org/

### CODE REVIEW

File: test/controllers/auth.test.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    mongoose
+      .connect(dbURI)
+      .then((result) => {
+        const payload = {
+          email: 'qa@test.com',
+          password: 'qatest123*',
+          name: 'QA',
+          posts: [],
+          _id: 'mock'
```
Comment (SuphawitCE) [inline test/controllers/auth.test.js]: For commit [3053ef9](https://github.com/SuphawitCE/node-restapi/pull/33/commits/3053ef97e6429e0f70fb51de96c92d9067528b88), need to change `_id` to be the same as in MongoDB to test duplicate user

File: test/controllers/auth.test.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      }
+    };
+
+    AuthController.getUserStatus(req, res, next).then(() => {
+      expect(res.statusCode).to.be.equal(200);
+      expect(req.userStatus).to.be.equal('New user');
+
+      done();
+    });
+  });
```
Comment (SuphawitCE) [inline test/controllers/auth.test.js]: At commit [38b7102](https://github.com/SuphawitCE/node-restapi/pull/33/commits/38b710290af0822ab95ee6b760a6da5e00ba632b), This case has a timeout issue

Screen Shot 2565-12-21 at 12 36 13

need to fix `done()`

**Comment by SuphawitCE** [comment]
At commit [aeab99a](https://github.com/SuphawitCE/node-restapi/pull/33/commits/aeab99ac2c2506dea4df3dd1f701c91d56d5abb0), build failing because the workflows can't run `npm test` and we won't public our credential to GitHub, so ignore this kind of error
